packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -895,6 +1080,687 @@ void Add(
+  const ColIndsArray& Ainds = A.graph.entries;
+  const ColIndsArray& Binds = B.graph.entries;
+  const RowPtrsArray* Arowptrs = (const RowPtrsArray*) &A.graph.row_map;
+  const RowPtrsArray* Browptrs = (const RowPtrsArray*) &B.graph.row_map;
+  GlobalColIndsArray AindsConverted("A colinds (converted)", Ainds.dimension_0());
+  GlobalColIndsArray BindsConverted("B colinds (converted)", Binds.dimension_0());
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+  auto MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("TpetraExt::MatrixMatrix::add() diff col map kernel: " + std::string("column map conversion"))));
+#endif
+  ConvertColIndsFunctor<GO, ColIndsArray, GlobalColIndsArray, local_map_type> convertA(minGlobalCol, Ainds, AindsConverted, AcolMap);
+  Kokkos::parallel_for(range_type(0, Ainds.dimension_0()), convertA);
+  ConvertColIndsFunctor<GO, ColIndsArray, GlobalColIndsArray, local_map_type> convertB(minGlobalCol, Binds, BindsConverted, BcolMap);
+  Kokkos::parallel_for(range_type(0, Binds.dimension_0()), convertB);
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+  MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("TpetraExt::MatrixMatrix::add() diff col map kernel: " + std::string("unsorted addition"))));
+#endif
+  UnsortedGlobalCols(
+     Avals, *Arowptrs, AindsConverted, scalarA,
+     Bvals, *Browptrs, BindsConverted, scalarB,
+     numGlobalCols, Cvals, Crowptrs, Ccolinds);
+}
+
+namespace AddUnsortedFunctors
+{
+//Non-ideal for peak memory usage but works and is completely parallel over rows (except for allocations and a few minor intermediate steps)
+//4 Functors:
+//  <Allocate CrowptrTemp>
+//1: compute C entries per row upper bound (A entries + B entries)
+//  <Make CrowptrTemp into prefix sum array>
+//  <Allocate oversized Cvals and Ccolind>
+//2: just copy scaled a,b values into oversized C values/colind (filling it completely)
+//  Note: these values aren't condensed or sorted yet (so, may be duplicate colinds, but that's expected)
+//  <Allocate Crowptr>
+//3: Sort each row's colinds (permuting values at same time), then count unique colinds (write that to Crowptr(i))
+//  <Make Crowptr into prefix sum array>
+//  <Allocate correctly sized Cvals and Ccolind>
+//4: Condense values from oversized Cvals/Ccolind to correctly sized
+//  <done: have everything needed to build final C>
+
+template<typename Ordinal, typename Values, typename RowPtrs, typename ColInds>
+struct EntryUpperBoundFunctor
+{
+  EntryUpperBoundFunctor(const Values& Avals_, const RowPtrs& Arowptrs_, const ColInds& Acolinds_, const Values& Bvals_, const RowPtrs& Browptrs_, const ColInds& Bcolinds_, RowPtrs& CrowcountsTemp_) :
+  Avals(Avals_), Arowptrs(Arowptrs_), Acolinds(Acolinds_),
+  Bvals(Bvals_), Browptrs(Browptrs_), Bcolinds(Bcolinds_),
+  CrowcountsTemp(CrowcountsTemp_) {}
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    CrowcountsTemp(i) = (Arowptrs(i + 1) - Arowptrs(i)) + (Browptrs(i + 1) - Browptrs(i));
+  }
+  const Values& Avals;
```

Comment:
See above note on Kokkos::View needing to be stored by value in functors, not by reference.